exporter/otlpexporter/otlp.go

```diff
@@ -133,62 +138,29 @@ func createOTLPExporter(config configmodels.Exporter) (*otlpExporter, error) {
 		}
 	}
 
-	numWorkers := defaultNumWorkers
-	if oCfg.NumWorkers > 0 {
-		numWorkers = oCfg.NumWorkers
+	exporter, serr := newExporter(oCfg)
+	if serr != nil {
+		return nil, fmt.Errorf("cannot configure OTLP exporter: %v", serr)
 	}
 
-	exportersChan := make(chan *exporterImp, numWorkers)
-	for exporterIndex := 0; exporterIndex < numWorkers; exporterIndex++ {
-		// TODO: newExporter blocks for connection. Now that we have ability
-		// to report errors asynchronously using Host.ReportFatalError we can move this
-		// code to Start() and do it in background to avoid blocking Collector startup
-		// as we do now.
-		exporter, serr := newExporter(oCfg)
-		if serr != nil {
-			return nil, fmt.Errorf("cannot configure OTLP exporter: %v", serr)
-		}
-		exportersChan <- exporter
-	}
-	oce := &otlpExporter{exporters: exportersChan}
+	oce := &otlpExporter{exporter: exporter}
 	return oce, nil
 }
 
 func (oce *otlpExporter) Shutdown(context.Context) error {
-	// Stop all exporters. Will wait until all are stopped.
-	wg := &sync.WaitGroup{}
-	var errors []error
-	var errorsMu sync.Mutex
-	visitedCnt := 0
-	for currExporter := range oce.exporters {
-		wg.Add(1)
-		go func(exporter *exporterImp) {
-			defer wg.Done()
-			err := exporter.stop()
-			if err != nil {
-				errorsMu.Lock()
-				errors = append(errors, err)
-				errorsMu.Unlock()
-			}
-		}(currExporter)
-		visitedCnt++
-		if visitedCnt == cap(oce.exporters) {
-			// Visited and concurrently executed stop() on all exporters.
-			break
-		}
-	}
-
-	// Wait for all stop() calls to finish.
-	wg.Wait()
-	close(oce.exporters)
-
-	return componenterror.CombineErrors(errors)
+	err := error(&exporterError{
+		code: errAlreadyStopped,
+		msg:  "OpenTelemetry exporter was already stopped.",
+	})
+	oce.stopOnce.Do(func() {
+		atomic.StoreInt32(&oce.state, stateStopped)
+		err = oce.exporter.stop()
+	})
+	return err
 }
 
 func (oce *otlpExporter) pushTraceData(ctx context.Context, td pdata.Traces) (int, error) {
-	// Get first available exporter.
-	exporter, ok := <-oce.exporters
-	if !ok {
+	if atomic.LoadInt32(&oce.state) != stateRunning {
```

Comment:
Supposedly this should not happen when used normally from the service.Application. We first shutdown receivers, then processors, then exporters so by the time exporters are shutdown there should be no more data in the pipeline.
Did you see this happening? (It can happen if called outside of service.Application).

Reply:
I tried to keep the previous behavior but probably not worth it, will remove.

Reply:
Removed, not necessary also none of the other exporters implement this.